File: leetcode/103. Binary Tree Zigzag Level Order Traversal/103.cpp
// https://helloacm.com/binary-tree-zigzag-level-order-traversal-algorithms-using-dfs-and-bfs/
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        dfs(root, 0);
        return res;
    }
private:
    vector<vector<int>> res;
    
    void dfs(TreeNode* root, int depth) {
        if (root == nullptr) return;
        if (res.size() <= depth) {
            res.push_back({});
        }
        if (depth % 2 == 0) {
            res[depth].push_back(root->val);
        } else {
            res[depth].insert(begin(res[depth]), root->val);
        }        
        dfs(root->left, depth + 1);
        dfs(root->right, depth + 1);        
    }
};

File: leetcode/103. Binary Tree Zigzag Level Order Traversal/103.py
# https://helloacm.com/teaching-kids-programming-binary-tree-zigzag-level-order-traversal-via-breadth-first-search-algorithm/
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
# MEDIUM, BFS

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        if not root:
            return ans
        q = deque([root])
        while q:
            n = len(q)
            cur = []
            for _ in range(n):
                x = q.popleft()
                cur.append(x.val)
                for kid in (x.left, x.right):
                    if kid:
                        q.append(kid)
            if len(ans) & 1:
                ans.append(cur[::-1])
            else:
                ans.append(cur)
        return ans
